### DISCUSSION TITLE
Serving html with templating?

### DISCUSSION THREAD

**Discussion by elibroftw** [matt-42/lithium#111] [General]
In flask, I have templates that flask templates with jinja2. I use the templating feature extensively in my web apps to avoid redundant code like the navigation bar and <head>.

Does lithium have something similar?

If not is there any c++ library that's maintained that comes with this feature I'm looking for?

**Comment by matt-42** [comment]
Hi @elibroftw 

Lithium does not provide a C++ template engine but there is definitely some templating libraries you can use with lithium. You can check this stackoverflow thread for candidates:
https://stackoverflow.com/questions/4456228/is-there-a-good-template-engine-for-c

**Reply by elibroftw** [reply]
So then for design should I memorize the html templates in memory and run the template engine on the string in memory or should I be reading the html file every request I need to respond to?

**Reply by matt-42** [reply]
If you need high performances, it's always better no to cache in memory yes

**Reply by elibroftw** [reply]
When I use li run is lithium using multiple cores and threads by default?

**Comment by jmafc** [comment]
I've been researching the same topic and FWIW found another SO question: https://stackoverflow.com/questions/355650/c-html-template-framework-templatizing-library-html-generator-library . I even tried Inja since it seemed the most promising (at least it has the most GH stars of `template-engine` in C++). My issue with it is that there are a lot of transformations. Assuming you have your data in some relational or other DB, you have to retrieve it, using SQL, an ORM or something similar, presumably into C++ classes/structs, but on the output side, Inja requires JSON-like inputs, so you have to convert your C++ objects into JSON-like objects. In Flask/Python, there is a much more seamless translation from DB to dicts/lists to what Jinja needs.

What I find surprising is that younger languages like Go and Rust have more HTML/text "template engines" than C++. Go has its own standard library implementation of templating, but there are competing libraries. And Rust has at least two "type-safe" engines that compile the templates directly into object code. Perhaps it's that in C++ "template" means generics/STL (even in SO questions people wonder what does the poster *really* want). Someone took this to the extreme: https://github.com/csb6/html-plus-plus.

Anyway, I'll be taking a look at Jinja2Cpp and perhaps some others. One other possibility is using std::format (in C++20) or fmt::format which does provide type-safe and even positional arguments (i.e., for i18n).
